tfimm/layers/lora.py

```diff
@@ -0,0 +1,92 @@
+import tensorflow as tf
+from dataclasses import dataclass
+from tfimm.architectures import convnext
+
+
+class LoraDense(tf.keras.layers.Dense):
+    def __init__(self, *args, lora_rank: int = 4, lora_alpha: float = 1, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.lora_rank = lora_rank
+        self.lora_alpha = lora_alpha
+        self.scaling = lora_alpha / lora_rank
+
+    def build(self, input_shape):
+        super().build(input_shape)
+        last_dim = self.kernel.shape[0]
+        self.kernel_0 = self.add_weight(
```

Comment:
I would not define `kernel_0`. We already get this weight, because we inherit from `Dense`. We simply define two additional weights. And then we override `call()` with our own implementation. Something like
```
def call(self, x):
    if self.inference_mode:  # Feature flag to run inference without speed penalty
        x = super().call(x)
    else:
        x1 = tf.matvec(self.kernel, x)
        x2 = tf.matvec(self.kernel_lora_b, x)
        x2 = tf.matvec(self.kernel_lora_a, x2)
        x = x1 + self.scale * x2 + self.bias
    return x
```
Side-note: Two matrix-vector multiplications should be faster than doing
```
x2 = tf.matvec(lora_a @ lora_b, x)
```
even though it is mathematically equivalent.

Reply:
**Regarding redefining existing weights:**

We can easily freeze non-LoRA layers using:
```python
for layer in model.layers:
    layer.trainable = False
```

But is there a way of freezing existing weights of a layer without redefining them?

As highlighted in this [issue](https://github.com/tensorflow/tensorflow/issues/47597), changing the trainable flag of specific weights is not a feature implemented in Keras.

Reply:
done